src/library.js

```diff
@@ -560,88 +560,74 @@ LibraryManager.library = {
+    //    Hence if an allocation fails, cut down on the amount of excess growth, in an attempt to succeed to perform a smaller allocation.
+    // According to these rules heap resize will occur at most 28 times throughout Wasm module lifetime.
+    // (first resize will resize up to 16MB, then 16MB*1.20^27 > 2048MB)
 
 #if WASM_MEM_MAX != -1
     // A limit was set for how much we can grow. We should not exceed that
-    // (the wasm binary specifies it, so if we tried, we'd fail anyhow). That is,
-    // if we are at say 64MB, and the max is 100MB, then we should *not* try to
-    // grow 64->128MB which is the default behavior (doubling), as 128MB will
-    // fail because of the max limit. Instead, we should only try to grow
-    // 64->100MB in this example, which has a chance of succeeding (but may
-    // still fail for another reason, of actually running out of memory).
-    newSize = Math.min(newSize, {{{ WASM_MEM_MAX }}});
-    if (newSize == oldSize) {
+    // (the wasm binary specifies it, so if we tried, we'd fail anyhow).
+    var maxHeapSize = {{{ WASM_MEM_MAX }}};
+#else
+    var maxHeapSize = 2147483648 - PAGE_MULTIPLE;
+#endif
+    if (requestedSize > maxHeapSize) {
 #if ASSERTIONS
-      err('Failed to grow the heap from ' + oldSize + ', as we reached the WASM_MEM_MAX limit (' + {{{ WASM_MEM_MAX }}} + ') set during compilation');
+      err('Cannot enlarge memory, asked to go up to ' + requestedSize + ' bytes, but the limit is ' + maxHeapSize + ' bytes!');
 #endif
       return false;
     }
-#endif // WASM_MEM_MAX
-
 #if USE_ASAN
-    // One byte of ASan's shadow memory shadows 8 bytes of real memory.
-    // If we increase the memory beyond 8 * ASAN_SHADOW_SIZE, then the shadow memory overflows.
-    // This causes real memory to be corrupted.
-    newSize = Math.min(newSize, {{{ 8 * ASAN_SHADOW_SIZE }}});
-    if (newSize == oldSize) {
+    // One byte of ASan's shadow memory shadows 8 bytes of real memory. Shadow memory area has a fixed size,
+    // so do not allow resizing past that limit.
+    maxHeapSize = Math.min(maxHeapSize, {{{ 8 * ASAN_SHADOW_SIZE }}});
+    if (requestedSize > maxHeapSize) {
 #if ASSERTIONS
       err('Failed to grow the heap from ' + oldSize + ', as we reached the limit of our shadow memory. Increase ASAN_SHADOW_SIZE.');
 #endif
       return false;
     }
 #endif
 
-    var replacement = emscripten_realloc_buffer(newSize);
-    if (!replacement) {
-#if ASSERTIONS
-      err('Failed to grow the heap from ' + oldSize + ' bytes to ' + newSize + ' bytes, not enough memory!');
+    var minHeapSize = 16777216;
+
+    // Loop through potential heap size increases. If we attempt a too eager reservation that fails, cut down on the
+    // attempted size and reserve a smaller bump instead. (max 3 times, chosen somewhat arbitrarily)
+    for(var cutDown = 1; cutDown <= 4; cutDown *= 2) {
+#if MEMORY_GROWTH_STEP == -1
+      var heapIncrement = oldSize * (1 + 0.20/cutDown); // ensure geometric growth
```

Comment:
this doesn't look like an increment, so the name is confusing I think - it's the new size, as it includes both the `1` factor plus the new incremental addition?

Reply:
renamed to `overGrownHeapSize`